pkg/operator/feature_gates.go

```diff
@@ -19,47 +19,113 @@ import (
 	"slices"
 	"strings"
 
-	k8sflag "k8s.io/component-base/cli/flag"
+	"github.com/prometheus/client_golang/prometheus"
 )
 
-// At the moment, the are no feature gates available.
-var defaultFeatureGates = map[string]bool{
-	"PrometheusAgentDaemonSet": false,
+const (
+	// PrometheusAgentDaemonSetFeature enables the DaemonSet mode for PrometheusAgent.
+	PrometheusAgentDaemonSetFeature FeatureGateName = "PrometheusAgentDaemonSet"
+)
+
+type FeatureGateName string
+
+type FeatureGates map[FeatureGateName]FeatureGate
+
+type FeatureGate struct {
+	description string
+	enabled     bool
 }
 
-// ValidateFeatureGates merges the feature gate default values with
+func (fg FeatureGates) Enabled(name FeatureGateName) bool {
+	return fg[name].enabled
+}
+
+// UpdateFeatureGates merges the current feature gate values with
 // the values provided by the user.
-func ValidateFeatureGates(flags *k8sflag.MapStringBool) (string, error) {
-	gates := defaultFeatureGates
-	if flags.Empty() {
-		return mapToString(gates), nil
+func (fg *FeatureGates) UpdateFeatureGates(flags map[string]bool) error {
+	for k := range flags {
+		f, found := (*fg)[FeatureGateName(k)]
+		if !found {
+			return fmt.Errorf("feature gate %q is unknown (supported feature gates: %s)", k, fg.String())
+		}
+		f.enabled = flags[k]
+		(*fg)[FeatureGateName(k)] = f
 	}
 
-	imgs := *flags.Map
-	for k, v := range imgs {
-		if _, ok := gates[k]; !ok {
-			return "", fmt.Errorf("feature gate %v is unknown", k)
-		}
-		gates[k] = v
+	return nil
+}
+
+func (fg *FeatureGates) keyValuePairs() ([]FeatureGateName, []FeatureGate) {
+	if fg == nil {
+		return nil, nil
+	}
+
+	var (
+		names = make([]FeatureGateName, 0, len(*fg))
+		gates = make([]FeatureGate, 0, len(*fg))
+	)
+	for k := range *fg {
+		names = append(names, k)
+	}
+	slices.Sort(names)
+
+	for _, v := range names {
+		gates = append(gates, (*fg)[v])
 	}
-	return mapToString(gates), nil
+
+	return names, gates
 }
 
-func AvailableFeatureGates() []string {
-	i := 0
-	gates := make([]string, len(defaultFeatureGates))
-	for k := range defaultFeatureGates {
-		gates[i] = k
-		i++
+func (fg *FeatureGates) Descriptions() []string {
+	var (
+		names, gates = fg.keyValuePairs()
+		desc         = make([]string, 0, len(names))
+	)
+
+	for i := range names {
+		desc = append(desc, fmt.Sprintf("%s: %s (enabled: %t)", names[i], gates[i].description, gates[i].enabled))
 	}
-	slices.Sort(gates)
-	return gates
+
+	return desc
 }
 
-func mapToString(m map[string]bool) string {
-	var s []string
-	for k, v := range m {
-		s = append(s, fmt.Sprintf("%s=%t", k, v))
+func (fg *FeatureGates) String() string {
+	names, gates := fg.keyValuePairs()
+
+	s := make([]string, len(names))
+	for i := range names {
+		s[i] = fmt.Sprintf("%s=%t", names[i], gates[i].enabled)
 	}
+
 	return strings.Join(s, ",")
 }
+
+var featureGateInfoDesc = prometheus.NewDesc(
+	"prometheus_operator_feature_gate_info",
```

Comment:
If it's a stateset now, ideally we'd drop the `_info` [suffix](https://github.com/OpenObservability/OpenMetrics/blob/main/specification/OpenMetrics.md#suffixes).

Reply:
yep, good catch!